src/librbd/image/RemoveRequest.cc

```diff
@@ -0,0 +1,641 @@
+    return nullptr;
+  }
+
+  cls::rbd::GroupSpec s;
+  bufferlist::iterator it = m_out_bl.begin();
+  *result = librbd::cls_client::image_get_group_finish(&it, &s);
+  if (*result < 0) {
+    send_close_image(*result);
+    return nullptr;
+  }
+  if (s.is_valid()) {
+    lderr(m_cct) << "image is in a consistency group - not removing" << dendl;
+    send_close_image(-EMLINK);
+    return nullptr;
+  }
+
+  if (!m_old_format) {
+    disable_features();
+  } else {
+    trim_image();
+  }
+
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::disable_features() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_disable_features>(this);
+
+  RWLock::RLocker owner_lock(m_image_ctx->owner_lock);
+  librbd::operation::DisableFeaturesRequest<I> *req =
+    librbd::operation::DisableFeaturesRequest<I>::create(
+      *m_image_ctx, ctx, 0, (m_image_ctx->features &
+                             (RBD_FEATURES_MUTABLE | RBD_FEATURES_DISABLE_ONLY)) &
+                             ~RBD_FEATURE_EXCLUSIVE_LOCK, true);
+  req->send();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_disable_features(int *result) {
+  ldout(m_cct, 20) << dendl;
+
+  if (*result < 0) {
+    send_close_image(*result);
+    return nullptr;
+  }
+
+  trim_image();
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::trim_image() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_trim_image>(this);
+
+  RWLock::RLocker owner_lock(m_image_ctx->owner_lock);
+  librbd::operation::TrimRequest<I> *req = librbd::operation::TrimRequest<I>::create(
+    *m_image_ctx, ctx, m_image_ctx->size, 0, m_prog_ctx);
+  req->send();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_trim_image(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "warning: failed to remove some object(s): " << cpp_strerror(*result)
+                 << dendl;
+  }
+
+  remove_child();
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::remove_child() {
+  ldout(m_cct, 20) << dendl;
+
+  m_image_ctx->parent_lock.get_read();
+  parent_info parent_info = m_image_ctx->parent_md;
+  m_image_ctx->parent_lock.put_read();
+
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::remove_child(&op, parent_info.spec, m_image_id);
+
+  using klass = RemoveRequest<I>;
+  librados::AioCompletion *rados_completion =
+    create_rados_ack_callback<klass, &klass::handle_remove_child>(this);
+  int r = m_image_ctx->md_ctx.aio_operate(RBD_CHILDREN, rados_completion, &op);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_remove_child(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    if (*result != -ENOENT) {
```

Comment:
Yes, -ENOENT is not an error, so we need to reset the error code for it. But what about the other ones? I mean, if we change it as below:
```
if (*result == -ENOENT)
      *result = 0;
```
Yes, that work. But the error message about other real error would be disappear. please take a look at the whole code.
```
  if (*result < 0) {
    if (*result != -ENOENT) {
      lderr(m_cct) << "error removing child from children list" << cpp_strerror(*result)
                   << dendl;
    } else {
      *result = 0;
    }
  }
```

The difference is the error message gone.